Sources/MIDIKit/MIDINetwork.swift

```diff
@@ -0,0 +1,199 @@
+//
+//  MIDINetwork.swift
+//  
+//
+//  Created by Michele Longhi on 28/10/2020.
+//
+
+import Foundation
+import CoreMIDI
+
+extension MIDIClient {
+    
+    public convenience init(name: String, networkConnection connection: MIDINetworkConnection) {
+        self.init(name: name)
+        
+        startWithNetworkConnection(connection)
+    }
+    
+    public func startWithNetworkConnection(_ connection: MIDINetworkConnection) {
+        
+        removeAllNetworkConnections()
+        
+        networkSession = MIDINetworkSession()
+        networkSession?.addConnection(connection)
+        
+        lastKnownConnection = connection
+        
+        try? start()
+    }
+
+    public func startWithFirstAvailableNetworkConnection(useLastKnownConnection: Bool = true, completion handler: ((_ connection: MIDINetworkConnection?)->Void)? = nil) {
+        
+        removeAllNetworkConnections()
+        
+        let networkSession = self.networkSession ?? MIDINetworkSession()
+        
+        if let connection = lastKnownConnection, useLastKnownConnection {
+            networkSession.addConnection(connection)
+            try? start()
+
+            handler?(connection)
+            return
+        }
+    
+        bonjourService.findService(BonjourService.Services.Apple_Midi, domain: BonjourService.LocalDomain) { services in
+            
+            if let service = services.first, let hostName = service.hostName {
+                
+                let connection = MIDINetworkConnection(host: .init(name: service.name, address: hostName, port: service.port))
+                networkSession.addConnection(connection)
+                try? self.start()
+
+                self.lastKnownConnection = connection
+                handler?(connection)
+            } else {
+                handler?(nil)
+            }
+        }
+    }
+    
+    public var networkSourceEndpoint: MIDIEndpoint? {
+        if let endpoint = networkSession?.sourceEndpoint() {
+            return MIDIEndpoint(endpoint)
+        }
+        return nil
+    }
+    
+    public var networkDestinationEndpoint: MIDIEndpoint? {
+        if let endpoint = networkSession?.destinationEndpoint() {
+            return MIDIEndpoint(endpoint)
+        }
+        return nil
+    }
+    
+    public func removeAllNetworkConnections() {
+        for connection in networkSession?.connections() ?? [] {
+            networkSession?.removeConnection(connection)
+        }
+    }
+    
+    public private(set) var lastKnownConnection: MIDINetworkConnection? {
+        get {
+            let ud = UserDefaults.standard
```

Comment:
Maybe make the `UserDefaults` container configureable and not always use the standard.

Reply:
Revisiting the last known connection persistence approach